text/0042-partial-fill-rules.md

```diff
@@ -0,0 +1,296 @@
+* In a simple transaction, if the sender intentionally sends an output to a recipient that has a bad masked_value, this is equivalent to sending money to into a hole. The recipient will not be able to recover the value, so they will not render services to the sender.
+* In MCIP #31 atomic swaps without partial fills, the originator is setting the masked_value for the required outputs which are going to them. If the counterparty changes the masked_value at all, the transaction is invalid. So the originator can only poison their own masked value, which only throws their own money into a hole.
+
+(This is also the reasoning for all other RingCT-based cryptocurrencies.)
+
+However, once we start doing partial fills, now we have a situation where the counter-party has to create the masked value for the originator. If consensus does not in some way validate the masked value, then a griefing attack becomes possible:
+
+* An originator makes a signed partial fill offer.
+* A counterparty fills the offer, but poisons the masked values going to the originator.
+* The counterparty walks away with their side of the exchange, and the originator technically received their fractional outputs, but view key matching fails to produce the correct value for these outputs, and so they cannot actually spend these outputs unless they brute-force the discrete logarithm over a u64 search space, which would be very expensive. (If you use baby-step giant-step algorithm here, then perhaps this can be done with ~4 billion elliptic curve operations. So, perhaps not impossible, but very inconvenient.)
+
+It is in principle possible that we could validate the masked value using a zero-knowledge proving framework. For example, in principle it is possible that a ZK-SNARK can prove that a particular Pedersen commitment and a particular masked_value both encrypt the same value, and this proof need not reveal what that value is.
+
+However, this is an extremely complicated solution to this problem. Embedding both Ristretto scalar multiplication and blake2b hashing into a ZK-SNARK would mean that we are building zero-knowledge proofs that are similar in complexity to Zcash. Making this mobile-friendly becomes a challenging research problem. So this is for now a dead-end and something we could only hope to do later as part of an upgrade to all of the transaction math.
+
+Another approach might involve migrating the masked_value from a simple symmetric-key encryption to something with homomorphic properties, so that we can verify it similarly to how we verify the partial-fill equations for the Pedersen commitments. However, this ultimately doesn't seem very promising, because even if we could homomorphically verify inequalities over the masked values, that isn't enough, because the masked value for each output has to actually match exactly to the corresponding Pedersen commitment, and two inequalities being satisfied doesn't imply this.
```

Comment:
:wrench: 
```suggestion
Another approach might involve migrating the masked value from a simple symmetric-key encryption to something with homomorphic properties, so that we can verify it similarly to how we verify the partial fill equations for the Pedersen commitments. However, this ultimately doesn't seem very promising, because even if we could homomorphically verify inequalities over the masked values, that isn't enough, because the masked value for each output has to actually match exactly to the corresponding Pedersen commitment, and two inequalities being satisfied doesn't imply this.
```